core/embed/rust/src/ui/component/marquee.rs

```diff
@@ -0,0 +1,229 @@
+        let now = Instant::now();
+
+        if let Event::Timer(token) = event {
+            if self.pause_token == Some(token) {
+                match self.state {
+                    State::PauseLeft => {
+                        let anim =
+                            Animation::new(self.max_offset, self.min_offset, self.duration, now);
+                        self.state = State::Right(anim);
+                    }
+                    State::PauseRight => {
+                        let anim =
+                            Animation::new(self.min_offset, self.max_offset, self.duration, now);
+                        self.state = State::Left(anim);
+                    }
+                    _ => {}
+                }
+                // We have something to paint, so request to be painted in the next pass.
+                ctx.request_paint();
+                // There is further progress in the animation, request an animation frame event.
+                ctx.request_anim_frame();
+            }
+
+            if token == EventCtx::ANIM_FRAME_TIMER {
+                if self.is_animating() {
+                    // We have something to paint, so request to be painted in the next pass.
+                    ctx.request_paint();
+                    // There is further progress in the animation, request an animation frame
+                    // event.
+                    ctx.request_anim_frame();
+                }
+
+                match self.state {
+                    State::Right(_) => {
+                        if self.is_at_right(now) {
+                            self.pause_token = Some(ctx.request_timer(self.pause));
+                            self.state = State::PauseRight;
+                        }
+                    }
+                    State::Left(_) => {
+                        if self.is_at_left(now) {
+                            self.pause_token = Some(ctx.request_timer(self.pause));
+                            self.state = State::PauseLeft;
+                        }
+                    }
+                    _ => {}
+                }
+            }
+        }
+        None
+    }
+
+    fn paint(&mut self) {
+        let now = Instant::now();
+
+        match self.state {
+            State::Initial => {
+                self.paint_anim(0);
+            }
+            State::PauseRight => {
+                self.paint_anim(self.min_offset);
+            }
+            State::PauseLeft => {
+                self.paint_anim(self.max_offset);
+            }
+            _ => {
+                let progress = self.progress(now);
+                if let Some(done) = progress {
+                    self.paint_anim(done as i16);
+                } else {
+                    self.paint_anim(0);
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "ui_debug")]
+impl crate::trace::Trace for Marquee {
+    fn trace(&self, d: &mut dyn crate::trace::Tracer) {
+        d.open("Marquee");
```

Comment:
This should also contain the text so we can match on it in tests.

Reply:
https://github.com/trezor/trezor-firmware/commit/2355d5755ccf6d652a2bd3d45b4d05caf3e097a8